chain/chain/src/stateless_validation/chunk_validation.rs

```diff
@@ -611,7 +637,21 @@ pub fn validate_chunk_state_witness(
     let (mut chunk_extra, mut outgoing_receipts) =
         match (pre_validation_output.main_transition_params, cache_result) {
             (MainTransition::Genesis { chunk_extra, .. }, _) => (chunk_extra, vec![]),
-            (MainTransition::NewChunk(new_chunk_data), None) => {
+            (MainTransition::NewChunk(mut new_chunk_data), None) => {
+                let mut validity_iterator =
+                    pre_validation_output.transaction_validity_check_passed.iter();
+                new_chunk_data.transactions.retain(|t| {
+                    let valid = *validity_iterator.next().unwrap();
+                    if !valid {
+                        tracing::debug!(
+                            target: "chain",
+                            message="discarding invalid transaction",
+                            tx=%t.get_hash()
+                        );
+                    }
+                    valid
+                });
+
                 let chunk_header = new_chunk_data.chunk_header.clone();
                 let NewChunkResult { apply_result: mut main_apply_result, .. } = apply_new_chunk(
                     ApplyChunkReason::ValidateChunkStateWitness,
```

Comment:
I have one more concern - this modifies the code for applying a chunk during state witness validation, but a chunk can also be applied in other ways. Nodes that track this shard will fetch the chunk and apply it independently, but I'm not sure if they'll filter the transactions in the same way. Applying a chunk with a different set of transactions could produce different results and the consensus could diverge.
Maybe it would be better to check this in `process_transactions` during chunk application?

Reply:
> Do you think it would be better to thread chain all the way to the bowels of the transaction runtime?

It'd be more foolproof, but I don't know how hard that is to do. We already have `EpochInfoProvider` in the runtime, maybe we could add a `ChainInfoProvider` as well?

Reply:
So, I experimented with this over the past week -- it is pretty darn difficult to implement something similar to `EpochInfoProvider`. First complicating factor is that in many cases with our current code structure the processing requires `: Send + Sync` or `'static` on this conceptual `ChainProvider`. To get the minimal required such `ChainProvider` be `'static`, we could put the `Chain::chain_store` into an `Arc` and move that around. But then this `Chain::chain_store` is accessed both mutably and immutably. So it also requires a `Mutex`.

From there a lot of our code is written assuming a) cheap access to `ChainStore`; b) to hold relatively long-lived references to `ChainStore`. So what would seem like initially a straightforward change suddenly becomes a big snowball rolling off a hillside.

I have a half-done prototype here: https://github.com/near/nearcore/compare/master...nagisa:nearcore:chain-provider-style-tx-validation with the main missing pieces being: a) updates to `ChainStoreUpdate` to work with `Arc<Mutex<ChainStore>>`; b) actual invocation of the newly added validation methods; c) some compile errors here and there, some `todo!()`s in places not related to state witness today which need additional threading through of data.

Reply:
I have updated the PR above to `cargo check` successfully. This means that I have threaded the `ChainProvider` through to the transaction runtime (`Runtime::apply`) from all callers.

There are three to five separate major call-points in regular code flow where transaction runtime has forced to thread `ChainProvider` through.

* `Chain::get_update_shard_job` -- internally validates transactions (calls `validate_chunk_transactions`;)
* stateless validation -- addressed here;
* `Chain::set_state_finalize` -- does not operate on untrusted chunks(?);
* catchup/state sync -- did not verify transaction validity before state witnesses and continues not validating them.

Out of these the catchup code smells most prone to problems here. Other than that everything else is in test code, utilities and tools.